### PR TITLE
Feature/grok support

### PR DESCRIPTION
**Pull request by mre**
This PR adds [grok](https://code.google.com/archive/p/semicomplete/wikis/Grok.wiki) support to gollum.

Grok allows parsing unstructured log-data using named regex patterns. Other log-multiplexers -  like [logstash](https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html) or [telegraf](https://github.com/influxdata/telegraf/blob/035905d65e13599c71f1c758e2e44e1184eef74d/plugins/inputs/logparser/README.md#grok-parser) - also offer grok support. Therefore adding this formatter would ease migration to gollum.

### CODE REVIEW

File: format/grok.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/trivago/gollum/core"
+	"github.com/vjeantet/grok"
+)
+
+// Grok formatter plugin
+// Grok is a formatter that applies regex filters on messages.
+// It works by combining text patterns into something that matches your logs.
+// Configuration example
+//
+//  Formatter: "format.Grok"
```
Comment (msiebeneicher) [inline format/grok.go]: Please use new config format here.

example: 

```
// myConsumer:
//   Type: consumer.File
//   File: /var/run/system.log
//   DefaultOffset: newest
//   OffsetFile: ""
//   Delimiter: "\n"
//   ObserveMode: poll
//   PollingDelay: 100
```

File: format/grok.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+//  Patterns:
+//    - (?<datacenter>[^\.]+?)\.(?<service>[^\.]+?)\.(?<host>[^\.]+?)\.(?<measurement>[^\s]+?)\s%{NUMBER:value:float}\s%{INT:time}
+//    - (?<datacenter>[^\.]+?)\.(?<service>[^\.]+?)\.(?<host>[^\.]+?)\.(?<application>twemproxy[^\.]+?)_(?<instance>[^\.]+?)\.(?<measurement>[^\s]+?)\s%{NUMBER:value:float}\s%{INT:time}
+//
+// The output format is JSON.
```
Comment (msiebeneicher) [inline format/grok.go]: Maybe than a better formatter name would be "grokToJSON"!?

File: format/grok.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+//   "measurement": "this.is.the.measurement",
+//   "value": "12.0",
+//   "time": "1497003802"
+// }
+// Patterns is a list of grok patterns that will be executed on the given message.
+// The first matching pattern will be used to parse the message.
+type Grok struct {
+	core.SimpleFormatter `gollumdoc:"embed_type"`
+	grok                 *grok.Grok
+	patterns             []string
```
Comment (msiebeneicher) [inline format/grok.go]: please use the new auto config here like

```golang
type Grok struct {
 +	core.SimpleFormatter `gollumdoc:"embed_type"`
 +	grok                 *grok.Grok
 +	patterns             []string `config:"Patterns"`
}
```

File: format/grok.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+func init() {
+	core.TypeRegistry.Register(Grok{})
+}
+
+// Configure initializes this formatter with values from a plugin config.
+func (format *Grok) Configure(conf core.PluginConfigReader) {
+	format.SimpleFormatter.Configure(conf)
+	format.patterns = conf.GetStringArray("Patterns", []string{})
```
Comment (msiebeneicher) [inline format/grok.go]: remove this two lines

File: format/grok.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+	grok, err := grok.NewWithConfig(&grok.Config{RemoveEmptyValues: true})
+
```
Comment (msiebeneicher) [inline format/grok.go]: remove empty line here

File: format/grok.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+// It returns a map of the defined values.
+func (format *Grok) applyGrok(content string) (map[string]string, error) {
+	for _, pattern := range format.patterns {
+		values, err := format.grok.Parse(pattern, content)
+
+		if err == nil && len(values) > 0 {
+			return values, nil
+		}
+	}
+	return nil, fmt.Errorf("Message does not match any grok pattern: %s", content)
```
Comment (msiebeneicher) [inline format/grok.go]: Please use the logger instead of 'fmt'

**Comment by msiebeneicher** [comment]
Please also add your new formatter to the readme

**Comment by msiebeneicher** [comment]
please reintegrate the master

**Comment by mre** [comment]
Reintegration done. I guess we can merge this? 😉

File: glide.lock [hunk 7]
```diff
@@ -131,8 +131,6 @@ imports:
   - tstrings
   - tsync
   - ttesting
-- name: github.com/vjeantet/grok
```
Comment (msiebeneicher) [inline glide.lock]: looks like you did not update the glide.lock properly after the reintegration.
try an update by `glide update github.com/vjeantet/grok --strip-vendor`

**Pull request merged**